File: frontend/src/types/watchlist.ts
// =======================================================
// Alpha India Watchlist Types
// Multi-Watchlist, Confidence Ranking (1-5), and Comments
// =======================================================

export interface WatchlistSummary {
  id: number;
  name: string;
  description: string;
  color: string;
  is_default: boolean;
  items_count: number;
  avg_confidence: number;
  created_at: string | null;
  updated_at: string | null;
}

export interface WatchlistItem {
  id: number;
  watchlist_id: number;
  symbol: string;
  company_name: string;
  confidence_score: number; // 1 to 5
  comment: string;
  target_price: number | null;
  added_at: string | null;
  updated_at: string | null;

  // Live Fundamentals
  current_price: number | null;
  sector: string;
  industry: string;
  exchange: string;
  market_cap: number | null;
  market_cap_category: string;
  stock_pe: number | null;
  roce: number | null;
  roe: number | null;
  sales_growth_3yr: number | null;
  profit_growth_3yr: number | null;
  return_3m: number | null;
  return_1y: number | null;
  dma_50: number | null;
  dma_200: number | null;
  ai_score: number | null;
}

export interface WatchlistDetailResponse {
  success: boolean;
  watchlist: WatchlistSummary;
  summary: {
    items_count: number;
    avg_confidence: number;
    avg_roce: number | null;
    high_conviction_count: number;
  };
  items: WatchlistItem[];
}

export interface StockSearchResult {
  symbol: string;
  company_name: string;
  sector: string;
  exchange: string;
  current_price: number | null;
  roce: number | null;
  stock_pe: number | null;
  market_cap: number | null;
}

export interface CreateWatchlistPayload {
  name: string;
  description?: string;
  color?: string;
}

export interface AddStockPayload {
  symbol: string;
  company_name?: string;
  confidence_score?: number;
  comment?: string;
  target_price?: number;
}

export interface UpdateStockPayload {
  confidence_score?: number;
  comment?: string;
  target_price?: number;
}

File: backend/app/models/watchlist.py
"""
Alpha India Watchlist Models
Supports multiple named watchlists, stock conviction/confidence ranking (1-5),
and inline investment thesis comments.
"""

from datetime import datetime
from sqlalchemy import (
    Column,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import relationship

from app.db.database import Base


class Watchlist(Base):
    __tablename__ = "watchlists"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=True)
    color = Column(String(30), default="cyan")  # cyan, emerald, amber, purple, blue
    is_default = Column(Integer, default=0)

    # Multi-tenant user ownership
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=True, index=True)

    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", backref="watchlists")
    items = relationship(
        "WatchlistItem",
        back_populates="watchlist",
        cascade="all, delete-orphan",
        order_by="WatchlistItem.confidence_score.desc(), WatchlistItem.id.asc()",
    )


class WatchlistItem(Base):
    __tablename__ = "watchlist_items"

    id = Column(Integer, primary_key=True, index=True)
    watchlist_id = Column(
        Integer,
        ForeignKey("watchlists.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    symbol = Column(String(30), nullable=False, index=True)
    company_name = Column(String(200), nullable=True)

    # Confidence Score: 1 to 5 (1 = Low/Speculative, 5 = Maximum Conviction)
    confidence_score = Column(Integer, default=3, nullable=False)

    # Investment Thesis / Note
    comment = Column(Text, nullable=True)

    # Target Price or Range
    target_price = Column(Float, nullable=True)

    added_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    watchlist = relationship("Watchlist", back_populates="items")

    # Composite Unique Constraint: A symbol cannot be added twice to the same watchlist
    __table_args__ = (
        UniqueConstraint("watchlist_id", "symbol", name="uq_watchlist_symbol"),
    )
